File: apps/ios/Loonext/Features/Settings/DeleteAccountCard.swift
import SwiftUI

/// What has to be typed. Fixed, unambiguous, and not a name we might change.
private let confirmWord = "delete"

/// #346 — deleting your own account, from the phone.
///
/// Apple 5.1.1(v) requires this of any app that lets you create an account,
/// but the reason to build it well is that a crew member who wants to leave
/// has had no way to: closing a workspace is the owner's alone, and being
/// removed by somebody else is not the same thing.
///
/// The copy draws the line the server actually draws. Someone deleting their
/// account will assume their texts to customers go with them; they do not,
/// they cannot (that record belongs to the business, and part of it sits under
/// a legal retention floor), and discovering it afterwards would be a
/// betrayal. So it is said before the button, not after.
@MainActor
struct DeleteAccountCard: View {
    let scope: SettingsScope
    let onDeleted: @MainActor () -> Void

    @State private var expanded = false
    @State private var preview: AccountDeletionPreview?
    @State private var loading = false
    @State private var error: String?
    @State private var confirming = false
    @State private var typed = ""
    @State private var deleting = false

    @Environment(\.appLocale) private var appLocale

    private var confirmed: Bool {
        typed.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() == confirmWord
    }

    private func t(_ key: String) -> String {
        AppStrings.translate(appLocale, key)
    }

    var body: some View {
        SettingsCard(
            title: t("settings.deleteTitle"),
            description: t("settings.deleteIntro")
        ) {
            if !expanded {
                Button(t("settings.deleteAction")) { load() }
                    .buttonStyle(.bordered)
                    .tint(BrandColor.destructive)
            } else if loading {
                Text(t("settings.deleteChecking"))
                    .font(.golos(12))
                    .foregroundStyle(BrandColor.muted600)
            } else if let error {
                InlineError(error)
            } else if preview?.blockedByOwnership == true {
                blockedByOwnership
            } else {
                consequences
            }
        }
        // The typed gate lives in the card, not in this alert: SwiftUI ignores
        // most modifiers on alert buttons, so a `.disabled` confirm there would
        // look gated and not be. The alert is the last plain "are you sure".
        .alert(t("settings.deleteConfirmTitle"), isPresented: $confirming) {
            Button(t("settings.deleteKeep"), role: .cancel) {}
            Button(t("settings.deleteAction"), role: .destructive) { remove() }
        } message: {
            Text(t("settings.deleteConfirmBody"))
        }
    }

    /// Specific, not generic: there is no ownership transfer yet (#332), so the
    /// way out has to be spelled out rather than discovered.
    /// "Brightside Plumbing" / "Brightside Plumbing, Rivera Roofing". Named
    /// rather than counted: the owner needs to know WHICH.
    private var ownedNames: String {
        let names = preview?.owned_workspaces.map(\.name) ?? []
        return names.isEmpty
            ? t("settings.deleteOwnedFallback")
            : names.joined(separator: ", ")
    }

    private var blockedByOwnership: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(AppStrings.translate(
                appLocale, "settings.deleteBlockedByOwnership", ["workspaces": ownedNames]
            ))
            .font(.golos(13))
            .foregroundStyle(BrandColor.ink)
            Text(t("settings.deleteClosingIsElsewhere"))
                .font(.golos(12))
                .foregroundStyle(BrandColor.muted600)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    private var consequences: some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(t("settings.deleteSignedOut"))
                .font(.golos(12))
                .foregroundStyle(BrandColor.muted600)

            if let preview, preview.memberships > 0 {
                Text(leavingSentence(preview))
                    .font(.golos(12))
                    .foregroundStyle(BrandColor.muted600)
            }

            Text(t("settings.deleteRecordStays"))
                .font(.golos(12))
                .foregroundStyle(BrandColor.muted600)

            // #371: said here rather than after the fact, because the moment
            // this succeeds you are signed out and there is no screen left to
            // read a confirmation on.
            Text(t("settings.deleteConfirmationEmail"))
                .font(.golos(12))
                .foregroundStyle(BrandColor.muted600)

            // The deliberate pause. This is the one place in the product where
            // slowing someone down is the right thing to do.
            TextField(
                AppStrings.translate(
                    appLocale, "settings.deleteTypeToConfirm", ["word": confirmWord]
                ),
                text: $typed
            )
                .textFieldStyle(.roundedBorder)
                .textInputAutocapitalization(.never)
                .autocorrectionDisabled()
                .font(.golos(13))
                .padding(.top, 6)

            Button(t("settings.deleteAction"), role: .destructive) { confirming = true }
                .buttonStyle(.borderedProminent)
                .tint(BrandColor.destructive)
                .disabled(!confirmed || deleting)
                .padding(.top, 6)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    /// Four whole sentences rather than a stem and a tail.
    ///
    /// The English reads as one sentence with a clause bolted on, and it cannot
    /// be translated that way: French agrees the verb with the count and puts
    /// the clause somewhere else. A shared stem would have printed half a
    /// sentence in one language and half in the other. Same split as Android's.
    private func leavingSentence(_ preview: AccountDeletionPreview) -> String {
        let openWork = preview.openWork > 0
        if preview.memberships == 1 {
            return t(
                openWork ? "settings.deleteLeaveOneOpenWork" : "settings.deleteLeaveOne"
            )
        }
        return AppStrings.translate(
            appLocale,
            openWork ? "settings.deleteLeaveManyOpenWork" : "settings.deleteLeaveMany",
            ["count": "\(preview.memberships)"]
        )
    }

    private func load() {
        expanded = true
        loading = true
        error = nil
        Task {
            do {
                preview = try await scope.repo.accountDeletionPreview()
            } catch {
                // `self.` is load-bearing: the caught `error` shadows the
                // property of the same name.
                self.error = t("settings.deletePreviewFailed")
            }
            loading = false
        }
    }

    private func remove() {
        deleting = true
        Task {
            do {
                _ = try await scope.repo.deleteAccount()
                onDeleted()
            } catch let failure {
                // `failure`, not `error` — the view already has an `error`
                // property and shadowing it here reads as a bug on every
                // future visit to this file.
                deleting = false
                typed = ""
                scope.showMessage(
                    (failure as? ApiError)?.message ?? t("settings.deleteFailed")
                )
            }
        }
    }
}

File: apps/android/app/src/main/kotlin/com/loonext/android/features/settings/DeleteAccountCard.kt
package com.loonext.android.features.settings

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.loonext.android.core.i18n.t
import kotlinx.coroutines.launch

/** What has to be typed. Fixed, unambiguous, and not a name we might change. */
private const val CONFIRM_WORD = "delete"

/**
 * #346 — deleting your own account, from the phone.
 *
 * Apple 5.1.1(v) and Play both require this, but the reason to build it well
 * is that a crew member who wants to leave has had no way to: closing a
 * workspace is the owner's alone, and being removed by somebody else is not
 * the same thing.
 *
 * The copy draws the line the server actually draws. Someone deleting their
 * account will assume their texts to customers go with them; they do not, they
 * cannot (that record belongs to the business, and part of it sits under a
 * legal retention floor), and discovering it afterwards would be a betrayal.
 * So it is said before the button, not after.
 */
@Composable
fun DeleteAccountCard(scope: SettingsScope, onDeleted: () -> Unit) {
    var expanded by remember { mutableStateOf(false) }
    var preview by remember { mutableStateOf<AccountDeletionPreview?>(null) }
    var loading by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    var confirming by remember { mutableStateOf(false) }
    var typed by remember { mutableStateOf("") }
    var deleting by remember { mutableStateOf(false) }
    val coroutines = rememberCoroutineScope()

    // Both failures are reported from a coroutine started by a press, where the
    // reader's locale is out of scope — so the sentences are read here.
    val previewFailed = t("settings.deletePreviewFailed")
    val deleteFailed = t("settings.deleteFailed")

    SettingsCard(
        title = t("settings.deleteTitle"),
        description = t("settings.deleteIntro"),
    ) {
        when {
            !expanded -> OutlinedButton(onClick = {
                expanded = true
                loading = true
                error = null
                coroutines.launch {
                    runCatching { scope.repo.accountDeletionPreview() }
                        .onSuccess { preview = it }
                        .onFailure { error = previewFailed }
                    loading = false
                }
            }) { Text(t("settings.deleteAction")) }

            loading -> Text(
                t("settings.deleteChecking"),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )

            error != null -> InlineError(error)

            preview?.blockedByOwnership == true -> Column {
                // Specific, not generic: there is no ownership transfer yet
                // (#332), so the way out has to be spelled out.
                Text(
                    t(
                        "settings.deleteBlockedByOwnership",
                        "workspaces" to preview!!.owned_workspaces.joinToString { it.name },
                    ),
                    style = MaterialTheme.typography.bodyMedium,
                )
                Spacer(Modifier.height(4.dp))
                Text(
                    t("settings.deleteClosingIsElsewhere"),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }

            else -> Column {
                Text(
                    t("settings.deleteSignedOut"),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                preview?.let { current ->
                    if (current.memberships > 0) {
                        Spacer(Modifier.height(6.dp))
                        Text(
                            // Four WHOLE sentences rather than a stem plus a
                            // clause. The English happens to read correctly when
                            // the two halves are concatenated; French moves the
                            // verb and the count, so a shared stem would leave a
                            // translator with two fragments neither of which can
                            // be made into a sentence.
                            when {
                                current.memberships == 1 && current.openWork > 0 ->
                                    t("settings.deleteLeaveOneOpenWork")

                                current.memberships == 1 -> t("settings.deleteLeaveOne")

                                current.openWork > 0 -> t(
                                    "settings.deleteLeaveManyOpenWork",
                                    "count" to current.memberships.toString(),
                                )

                                else -> t(
                                    "settings.deleteLeaveMany",
                                    "count" to current.memberships.toString(),
                                )
                            },
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                }
                Spacer(Modifier.height(6.dp))
                Text(
                    t("settings.deleteRecordStays"),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                Spacer(Modifier.height(6.dp))
                // #371: said here rather than after the fact, because the
                // moment this succeeds you are signed out and there is no
                // screen left to read a confirmation on.
                Text(
                    t("settings.deleteConfirmationEmail"),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                Spacer(Modifier.height(12.dp))
                Button(
                    onClick = { confirming = true },
                    colors = ButtonDefaults.buttonColors(
                        containerColor = MaterialTheme.colorScheme.error,
                        contentColor = MaterialTheme.colorScheme.onError,
                    ),
                ) { Text(t("settings.deleteAction")) }
            }
        }
    }

    if (confirming) {
        AlertDialog(
            onDismissRequest = { if (!deleting) { confirming = false; typed = "" } },
            title = { Text(t("settings.deleteConfirmTitle")) },
            text = {
                Column {
                    Text(
                        t("settings.deleteConfirmBody"),
                        style = MaterialTheme.typography.bodyMedium,
                    )
                    Spacer(Modifier.height(12.dp))
                    OutlinedTextField(
                        value = typed,
                        onValueChange = { typed = it },
                        singleLine = true,
                        enabled = !deleting,
                        modifier = Modifier.fillMaxWidth(),
                        label = { Text(t("settings.deleteTypeToConfirm", "word" to CONFIRM_WORD)) },
                    )
                }
            },
            confirmButton = {
                LinkButton(
                    enabled = !deleting && typed.trim().equals(CONFIRM_WORD, ignoreCase = true),
                    onClick = {
                        deleting = true
                        coroutines.launch {
                            runCatching { scope.repo.deleteAccount() }
                                .onSuccess {
                                    confirming = false
                                    onDeleted()
                                }
                                .onFailure { cause ->
                                    deleting = false
                                    confirming = false
                                    typed = ""
                                    scope.showMessage(cause.message ?: deleteFailed)
                                }
                        }
                    },
                ) {
                    Text(
                        if (deleting) t("settings.deletePending") else t("settings.deleteAction"),
                    )
                }
            },
            dismissButton = {
                LinkButton(
                    enabled = !deleting,
                    onClick = { confirming = false; typed = "" },
                ) { Text(t("settings.deleteKeep")) }
            },
            modifier = Modifier.padding(0.dp),
        )
    }
}
